dbeaver/dbeaver#1017 Add different ERD diagram types

serjiokov opened the pull request:
Provided osgi service (registry), extension point definition, objecy model descriptor for contribution ERD notation styles.
issue: https://github.com/dbeaver/dbeaver/issues/1017

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Prepared service, extension point definition and contribution of ERD
notation IDEF1X.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Prepared default contribution for ERD notation - IDEF1X.

serge-rider commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/internal/notation/ERDNotationServiceImpl.java:
```diff
@@ -0,0 +1,49 @@
+package org.jkiss.dbeaver.erd.ui.internal.notation;
```
Copyright

serjiokov replied:
Done

serge-rider commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationService.java:
```diff
@@ -0,0 +1,31 @@
+
+import java.util.Collection;
+
+public interface ERDNotationService<T extends ERDNotationDescriptor> {
```
Let's use our classic approach with registries (and no osgi services).
Similarly to ERDExportFormatRegistry.
Otherwise we mix different approaches even in the single bundle, it produces a mess in the codebase.

serjiokov replied:
Reworked on registry approach.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Service reorganized to singleton registry.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Added new ERD notations implementation: IDEF1X and BachMan.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Renamed stubs contributions to required.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Highlighted connection arrow

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Crows foot notation implementation (part 1)

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Provided integration notations for editor.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fix preferences listener to be responsive on user changes.

serjiokov commented:
Provided implementation of representation relational notations: IDEF1X (default), Crows Foot, BachMan. User have chance to change notation through pop-up menu.
1) EDEF1X (default)
![image](https://github.com/dbeaver/dbeaver/assets/32031248/6d7a81b0-fa90-44f1-b96d-a6e778acc074)
2) Crows Foot
![image](https://github.com/dbeaver/dbeaver/assets/32031248/755958ad-a7bb-46fa-9b5c-171f4a92aa0d)
3) Bachman
![image](https://github.com/dbeaver/dbeaver/assets/32031248/68d87def-0507-49fc-90a4-1baca53a81e5)

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Unused import - org.jkiss.dbeaver.erd.ui.notations.ERDNotationRegistry.

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/plugin.xml:
```diff
@@ -590,5 +591,29 @@
     </extension>
+    <extension
+          point="org.jkiss.dbeaver.erd.ui.notation.style">
+       <erdnotaion
+             description="Integration DEFinition for information modeling (IDEF1X) notation"
```
Typo in name

serjiokov replied:
Fixed

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:
```diff
@@ -1079,6 +1089,29 @@ public void run()
+        @Override
+        public void run() {
+            getDiagram().setDiagramNotation(notation);
+            refreshDiagram(true, true);
```
Can we avoid refreshing the diagram layout, or is it needed? We can update entities and relations separately.

serjiokov replied:
Please point me how to update relations, but in general for further cases require to refresh diagram entierly.

E1izabeth replied:
@Destrolaric ?

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDPreferencePage.java:
```diff
@@ -144,19 +163,22 @@ private void createStyleGroup(DBPPreferenceStore store, Composite composite)
+            store.getBoolean(ERDUIConstants.PREF_GRID_SNAP_ENABLED), 2);
+        spinnerGridWidth = UIUtils.createLabelSpinner(gridGroup, ERDUIMessages.pref_page_erd_spinner_grid_width,
+            store.getInt(ERDUIConstants.PREF_GRID_WIDTH), 5, Short.MAX_VALUE);
+        spinnerGridHeight = UIUtils.createLabelSpinner(gridGroup, ERDUIMessages.pref_page_erd_spinner_grid_height,
+            store.getInt(ERDUIConstants.PREF_GRID_HEIGHT), 5, Short.MAX_VALUE);
```
Are you sure that we want such a range for the spinner? Short.MAX_VALUE is too big by my opinion

serjiokov replied:
Previous code, just formated to avoid code style issues

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDPreferencePage.java:
```diff
@@ -128,11 +148,10 @@ private void createVisibilityGroup(DBPPreferenceStore store, Composite composite
         ERDViewStyle[] enabledStyles = ERDViewStyle.getDefaultStyles(store);
 
-        Group elemsGroup = UIUtils.createControlGroup(composite, ERDUIMessages.erd_preference_page_title_attribute_style, 1, GridData.VERTICAL_ALIGN_BEGINNING | GridData.FILL_HORIZONTAL, 0);
+        Group elemsGroup = UIUtils.createControlGroup(composite, ERDUIMessages.erd_preference_page_title_attribute_style, 1, GridData.FILL_BOTH, 0);
```
Could you attach the new preference page screenshot?

serjiokov replied:
![image](https://github.com/dbeaver/dbeaver/assets/32031248/b1d4fdba-0403-46d4-a519-6f4b087b59fd)

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/model/EntityDiagram.java:
```diff
@@ -200,4 +205,13 @@ public boolean isDirty() {
+        return notationStyle;
+    }
+    
+    public void setDiagramNotation(ERDNotationDescriptor notation) {
+        this.notationStyle = notation;
+    }
```
Could you cover methods with annotations, it will help us later to understand the nullability of values

serjiokov replied:
Done

ShadelessFox replied:
I don't see any changes.

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notation/crowsfoot/CrowsFootDiagramNotation.java:
```diff
@@ -0,0 +1,86 @@
+                conn.setTargetDecoration(targetDecor);
+            }
+        }
+        conn.setLineWidth(1);
+        conn.setLineStyle(SWT.LINE_CUSTOM);
```
Also, provide screenshots with the shortest path and orthogonal routing.

serjiokov replied:
![image](https://github.com/dbeaver/dbeaver/assets/32031248/60912837-61b0-48c4-9cdf-13a8c11e480d)

Destrolaric commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notation/crowsfoot/CrowsFootPolylineDecoration.java:
```diff
@@ -0,0 +1,96 @@
+
+        PointList crowsFootPoints = getPoints();
+        Point point;
+        int radius = DIAMETR / 2;
```
typo

serjiokov replied:
Fixed.

Destrolaric reviewed (changes requested):
Could you also look at the CI errors, there are some formatting issues

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed code for IDEA style

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed types

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed code style requirement issue

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed code style requirement issue
Javadoc tag @param should be preceded with an empty line.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed code style requirement issue
Javadoc tag @param should be preceded with an empty line.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:
```diff
@@ -1079,6 +1095,29 @@ public void run()
+
+        @Override
+        public boolean isChecked() {
+            if (notation == null || getDiagram().getDiagramNotation() == null) {
```
The `notation` field cannot be null. That's why we use `@NotNull`/`@Nullable` annotations.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/editor/ERDEditorPart.java:
```diff
@@ -1079,6 +1095,29 @@ public void run()
+    private class ChangeERDNotationStyleAction extends Action {
+        private final ERDNotationDescriptor notation;
+
+        public ChangeERDNotationStyleAction(ERDNotationDescriptor notation) {
```
```suggestion
        public ChangeERDNotationStyleAction(@NotNull ERDNotationDescriptor notation) {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationRegistry.java:
```diff
@@ -0,0 +1,116 @@
+     *@param name - notation name
+     *@return - Optional ERDNotationDescriptor
+     */
+    public Optional<ERDNotationDescriptor> getERDNotationByName(String name) {
```
Uh, no. Don't ever use `Optional` outside Stream APIs.

Return `ERDNotationDescriptor`, and annotate it with `@Nullable`.

serjiokov replied:
Please argue "`Uh, no. Don't ever use Optional outside Stream API" `

serjiokov replied:
PLS read this: https://ducmanhphan.github.io/2019-12-06-Best-practice-for-Optional-in-Java/

ShadelessFox replied:
We don't use `Optional` in our code, especially in getters. A nullable reference is enough. Don't refer to other resources.

E1izabeth replied:
Let's discuss Optional usage in a separate meeting. I wrote that point down.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationRegistry.java:
```diff
@@ -0,0 +1,116 @@
+     *@param id - notation descriptor identifier
+     *@return - ERDNotationDescriptor
+     */
+    public ERDNotationDescriptor getNotation(String id) {
```
Please specify nullability annotations for the method's parameters and its return type.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationRegistry.java:
```diff
@@ -0,0 +1,116 @@
+        return notations.get(id);
+    }
+
+    public ERDNotationDescriptor getDefaultNotation() {
```
And here

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java:
```diff
@@ -57,7 +57,7 @@
  */
 public class AssociationPart extends PropertyAwareConnectionPart {
-
+    Log log = Log.getLog(this.getClass());
```
```suggestion
    private static final Log log = Log.getLog(AssociationPart.class);
```

E1izabeth replied:
I agree that it's better to at least add `private static final` modifiers.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/part/AssociationPart.java:
```diff
@@ -164,46 +164,14 @@ protected void setConnectionRouting(PolylineConnection conn) {
-            }
-            conn.setLineDash(
-                constraintType.isLogical() ? new float[]{ 4  } : new float[]{ 5 });
+        if (diagramNotation != null) {
```
```suggestion
        else {
```

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notation/crowsfoot/CrowsFootPolylineDecoration.java:
```diff
@@ -0,0 +1,95 @@
+        switch (association) {
+            case ZERO_OR_ONE:
+                g.drawLine(crowsFootPoints.getPoint(3), crowsFootPoints.getPoint(4));
+                point = crowsFootPoints.getPoint(7);
```
Please extract this assignment from all paths to just a single place

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDAssociationType.java:
```diff
@@ -0,0 +1,26 @@
+public enum ERDAssociationType {
+    ZERO,
+    ZERO_OR_ONE,
+    ONE_ONLY,
```
As we already have `ZERO`, let it be called similarly without the `ONLY` prefix
```suggestion
    ONE,
```

serjiokov replied:
Don't think so, prefer to keep all notations defined in specification, ONE and ONLY_ONE is different take a look on screens:
![Screenshot 2023-09-20 085534](https://github.com/dbeaver/dbeaver/assets/32031248/825a66a4-ff3d-42c4-a178-2258a34379eb)
![Screenshot 2023-09-20 085433](https://github.com/dbeaver/dbeaver/assets/32031248/f70cd4b1-63a5-466f-aa71-56529172194d)

E1izabeth replied:
Yeah, we might want to add ONE later. Currently we can't check uniqueness to build such associations, however, we have an old issue about supporting/checking/showing unique constraints in erd.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotation.java:
```diff
@@ -0,0 +1,46 @@
+     *@param bckColor    - back end color
+     *@param frgColor    - front end color
+     */
+    public void applyNotationForEntities(PolylineConnection conn, ERDAssociation association, Color bckColor, Color frgColor);
```
Please use complete names instead of abbreviations in interfaces and annotate parameters with nullability annotations as it's not clear what arguments are supposed to be provided.

- `conn` -> `connection`
- `bckColor` -> `backgroundColor` (or `background`)
- `frgColor` -> `foregroundColor` (or `foreground`)

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationBase.java:
```diff
@@ -0,0 +1,27 @@
+
+package org.jkiss.dbeaver.erd.ui.notations;
+
+public class ERDNotationBase {
```
Sounds more like `ERDNotationConstants`

serjiokov replied:
The base class name more generic and created specifically for further step, naming class with constant prefix is not a require.

ShadelessFox replied:
Then better make it `abstract` and add `implements ERDNotation`, or just convert it into an interface but still extend the `ERDNotation` interface to mark it as a base type for other notations.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationDescriptor.java:
```diff
@@ -0,0 +1,66 @@
+    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_NAME = "name"; // $NON-NLS-N$
+    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DESCRIPTION = "description"; // $NON-NLS-N$
+    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DEFAULT = "isDefault"; // $NON-NLS-N$
+    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_NOTATION = "notation"; // $NON-NLS-N$
```
Why are these public? They are only used in a single place. Mark them as `public` and remove redundant prefixes. We also already have some of these predefined in the `RegistryConstants` class. Reuse if possible.

```suggestion
    public static final String ATTR_ID = "id"; // $NON-NLS-N$
    public static final String ATTR_NAME = "name"; // $NON-NLS-N$
    public static final String ATTR_DESCRIPTION = "description"; // $NON-NLS-N$
    public static final String ATTR_DEFAULT = "isDefault"; // $NON-NLS-N$
    public static final String ATTR_NOTATION = "notation"; // $NON-NLS-N$
```

I'd also like the `notation` attribute to be renamed to `class` as it's the implementation class.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationDescriptor.java:
```diff
@@ -0,0 +1,66 @@
+        this.name = cf.getAttribute(ERD_STYLE_NOTATION_EXT_ATTRIBUTE_NAME);
+        this.description = cf.getAttribute(ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DESCRIPTION);
+        this.isDefault = Boolean.valueOf(cf.getAttribute(ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DEFAULT));
+        this.notation = (ERDNotation) cf.createExecutableExtension(ERD_STYLE_NOTATION_EXT_ATTRIBUTE_NOTATION);
```
Please use the `ObjectType` wrapper for lazy initialization. Look at other descriptors for a reference.

serjiokov replied:
Lazy initialization is not require.

ShadelessFox replied:
We always use lazy initialization in descriptors in such cases.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationRegistry.java:
```diff
@@ -0,0 +1,116 @@
+
+    private ERDNotationRegistry(IExtensionRegistry registry) {
+        IConfigurationElement[] cfgElements = registry.getConfigurationElementsFor(EXTENSION_ID);
+        Stream.of(cfgElements).forEach(c -> {
```
Stop abusing Stream APIs

```suggestion
        for (IConfigurationElement c : cfgElements) {
```

serjiokov replied:
Argue your comment pls : `"Stop abusing Stream APIs"  `

ShadelessFox replied:
Streams create unnecessary noise and slow down code. What's the point of creating a new stream and iterating over its elements rather than using a plain for loop?

serjiokov replied:
The java Stream does not store elements and has a functional in nature, this is standard for since Java 8 been released.
A lot of greate articles covering Streams vs For, one of them:
https://blog.knoldus.com/java-streams-vs-loop/

ShadelessFox replied:
Streams are good for a processing pipeline. You filter objects, map them, filter again, sort and collect. If you were to do that using for loops, it would require way more code than using streams. But if we are talking about just iterating over elements of an array, streams are just too much. You instantiate a new Stream object, create a bootstrap method with a callback for the `forEach` method, and then, under the hood, it performs a plain `foreach` loop under the hood, but slower and requiring more space for the compiled code on the disk. What's the point?

serjiokov replied:
Ok, I vote for Streams anyway, I don't think that using Stream API increase compiled code base significally. Using Streams is a way to reach more efficient code, that looks greate and simplify product maintance.

ShadelessFox replied:
> Using Streams is a way to reach more efficient code, that looks greate and simplify product maintenance

Not in **that particular** case.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationRegistry.java:
```diff
@@ -0,0 +1,116 @@
+     *
+     *@param descriptor - notation descriptor
+     */
+    public void addNotation(@NotNull ERDNotationDescriptor descriptor) {
```
This should be a `private` method.

serjiokov replied:
Done

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Fixed by comments

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationDescriptor.java:
```diff
@@ -22,11 +22,11 @@
-    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DESCRIPTION = "description"; // $NON-NLS-N$
-    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_DEFAULT = "isDefault"; // $NON-NLS-N$
-    public static final String ERD_STYLE_NOTATION_EXT_ATTRIBUTE_NOTATION = "notation"; // $NON-NLS-N$
+    private static final String EXT_ATTRIBUTE_ID = "id"; // $NON-NLS-N$
+    private static final String EXT_ATTRIBUTE_NAME = "name"; // $NON-NLS-N$
+    private static final String EXT_ATTRIBUTE_DESCRIPTION = "description"; // $NON-NLS-N$
+    private static final String EXT_ATTRIBUTE_DEFAULT = "isDefault"; // $NON-NLS-N$
+    private static final String EXT_ATTRIBUTE_NOTATION = "notation"; // $NON-NLS-N$
```
Please reuse constants from the `RegistryConstants` class.

serjiokov replied:
What idea behind the RegistryConstants?  all words for all times?

ShadelessFox replied:
It's just a place for all common constants used in extension points. It keeps the code clean by limiting the number of duplicated constants scattered across the code base.

serjiokov replied:
It is not true, this class - RegistryConstants keeps generic and specific parts of each extension point, that should be part of descriptor.

ShadelessFox replied:
Sorry, I don't understand what you are saying.

serjiokov replied:
Your point `It's just a place for all common constants used in extension points.`
This class RegistryConstants keeps both generic and specific parts for each extension point, specific attributes is not a common.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotation.java:
```diff
@@ -0,0 +1,46 @@
+     * @param bckColor    - back end color
+     * @param frgColor    - front end color
+     */
+    public void applyNotationForArrows(PolylineConnection conn, ERDAssociation association, Color bckColor, Color frgColor);
+
+    /**
+     * Method designed to display diagram entity notation
+     *
+     * @param conn        - connection class specification
+     * @param association - ERD association
+     * @param bckColor    - back end color
+     * @param frgColor    - front end color
+     */
+    public void applyNotationForEntities(PolylineConnection conn, ERDAssociation association, Color bckColor, Color frgColor);
```
I don't see any annotations added here.

ShadelessFox commented on plugins/org.jkiss.dbeaver.erd.ui/src/org/jkiss/dbeaver/erd/ui/notations/ERDNotationDescriptor.java:
```diff
@@ -0,0 +1,66 @@
+        this.name = cf.getAttribute(EXT_ATTRIBUTE_NAME);
+        this.description = cf.getAttribute(EXT_ATTRIBUTE_DESCRIPTION);
+        this.isDefault = Boolean.valueOf(cf.getAttribute(EXT_ATTRIBUTE_DEFAULT));
+        this.notation = (ERDNotation) cf.createExecutableExtension(EXT_ATTRIBUTE_NOTATION);
```
Please use `ObjectType` instead of directly initializing it here.

serjiokov replied:
Please see the example implementation from parallel class
```
 public ERDExportFormatHandler getInstance() throws DBException {
            return type.createInstance(ERDExportFormatHandler.class);
        }
```
During the first call of registry (getInstance()) the instance of extension class created, no Lazy instantiating here.

serjiokov replied:
Added Lazy wrapper usage.

serjiokov committed:
dbeaver/dbeaver#1017 Add different ER-diagram notations

Lazy wrapper for lazy registry

serjiokov committed:
dbeaver/dbeaver#1017 Code style requirement

serjiokov committed:
dbeaver/dbeaver#1017 Added abstract class, back Stream to Loops

serjiokov committed:
dbeaver/dbeaver#1017 Code style check

serjiokov committed:
dbeaver/dbeaver#1017 NotNullabe and Nullable

serjiokov committed:
dbeaver/dbeaver#1017 Javadoc tag @return should be preceded with an
empty line.

serjiokov committed:
dbeaver/dbeaver#1017 Removed Streams and Optional usages

Matvey16 commented:
Verified

E1izabeth committed:
Merge branch 'devel' into feature#1017-add-different-er-diagram-types
